File: iOS-userLocation-smoothMove/ViewController.m
//
//  ViewController.m
//  iOS-userLocation-smoothMove
//
//  Created by shaobin on 16/12/22.
//  Copyright © 2016年 autonavi. All rights reserved.
//

#import "ViewController.h"
#import <MAMapKit/MAMapKit.h>
#import "MovingAnimation.h"

@interface ViewController ()<MAMapViewDelegate>

@property (nonatomic, strong) MAMapView *mapView;
@property (nonatomic, strong) MAPointAnnotation *customUserAnnotation;
@property (nonatomic, strong) MovingAnimation *movingAnimation;
@property (nonatomic, strong) NSTimer *timer;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.mapView = [[MAMapView alloc] initWithFrame:self.view.bounds];
    self.mapView.customizeUserLocationAccuracyCircleRepresentation = YES;
    self.mapView.delegate = self;
    [self.view addSubview:self.mapView];
    
    self.customUserAnnotation = [[MAPointAnnotation alloc] init];
    self.movingAnimation = [[MovingAnimation alloc] init];
    
    self.mapView.zoomLevel = 16;
    
    [self initBtn];
    
}

- (void)dealloc {
    [self stopTimer];
}

- (void)initBtn {
    
    UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(10, 100, 120, 32)];
    button.backgroundColor = [UIColor whiteColor];
    button.layer.borderWidth = 1.0;
    button.titleLabel.font = [UIFont systemFontOfSize:12];
    button.layer.borderColor = [UIColor redColor].CGColor;
    [button setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    [button setTitle:@"showsUserlocation" forState:UIControlStateNormal];
    
    [button addTarget:self action:@selector(actionLocation) forControlEvents:UIControlEventTouchUpInside];
    
    [button sizeToFit];
    [self.view addSubview:button];
}

- (void)actionLocation {
    self.mapView.showsUserLocation = !self.mapView.showsUserLocation;
    
    if (!self.mapView.showsUserLocation) {
        [self.mapView removeAnnotation:self.customUserAnnotation];
    }
}

#pragma mark - MAMapViewDelegate

- (void)mapView:(MAMapView *)mapView didUpdateUserLocation:(MAUserLocation *)userLocation updatingLocation:(BOOL)updatingLocation {
    if(!updatingLocation || !userLocation.location) {
        return;
    }
    
    if([self.mapView.annotations containsObject:self.customUserAnnotation]) {
        CLLocationCoordinate2D preCoord = self.customUserAnnotation.coordinate;
        CLLocationCoordinate2D curCoord = userLocation.location.coordinate;
        
        
        /********************************
         模拟定位飘了的情况，方便测试用的
         ********************************/
        static int simulatedCount = 0;
        if(simulatedCount < 10) {
            float delta = 0.001 + (float)(rand() % 100) / 100000;
            float delta2 = 0.001 + (float)(rand() % 100) / 100000;
            delta *= (rand() % 2 == 0 ? 1 : -1);
            delta2 *= (rand() % 2 == 0 ? 1 : -1);
            curCoord.latitude = preCoord.latitude + delta;
            curCoord.longitude = (float)preCoord.longitude + delta2;
            
            simulatedCount++;
        }
        
        CLLocation *prevLoc = [[CLLocation alloc] initWithLatitude:preCoord.latitude longitude:preCoord.longitude];
        CLLocation *loc = [[CLLocation alloc] initWithLatitude:curCoord.latitude longitude:curCoord.longitude];
        
        CGPoint prevPoint = [self.mapView convertCoordinate:preCoord toPointToView:self.mapView];
        CGPoint curPoint = [self.mapView convertCoordinate:curCoord toPointToView:self.mapView];
        
        //超过两个像素才做动画
        if(fabs(curPoint.x - prevPoint.x) > 2 || fabs(curPoint.y - prevPoint.y) > 2) {
            [self addMovingAnnotationFrom:prevLoc toCoord:loc];
        } else {
            self.customUserAnnotation.coordinate = curCoord;
        }
    } else {
        [self.mapView addAnnotation:self.customUserAnnotation];
        self.customUserAnnotation.coordinate = userLocation.location.coordinate;
        self.mapView.centerCoordinate = userLocation.location.coordinate;
    }
}

- (MAOverlayRenderer *)mapView:(MAMapView *)mapView rendererForOverlay:(id<MAOverlay>)overlay {
    if(overlay == self.mapView.userLocationAccuracyCircle) {
        return nil;
    }
    
    return nil;
}

- (MAAnnotationView *)mapView:(MAMapView *)mapView viewForAnnotation:(id<MAAnnotation>)annotation {
    if([annotation isKindOfClass:[MAUserLocation class]]) {
        NSString *reuseId = @"defaultUserAnnotation";
        MAAnnotationView * v = [self.mapView dequeueReusableAnnotationViewWithIdentifier:reuseId];
        if(v) {
            return v;
        }
        MAAnnotationView *userLocationView = [[MAAnnotationView alloc] initWithAnnotation:self.customUserAnnotation reuseIdentifier:reuseId];
        userLocationView.frame = CGRectZero;
        return userLocationView;
    }
    
    if(annotation == self.customUserAnnotation) {
        NSString *reuseId = @"customUserAnnotation";
        MAAnnotationView * v = [self.mapView dequeueReusableAnnotationViewWithIdentifier:reuseId];
        if(v) {
            return v;
        }
        MAAnnotationView *userLocationView = [[MAAnnotationView alloc] initWithAnnotation:self.customUserAnnotation reuseIdentifier:reuseId];
        userLocationView.frame = CGRectMake(0, 0, 20, 20);
        userLocationView.backgroundColor = [[UIColor blueColor] colorWithAlphaComponent:0.8];
        userLocationView.layer.cornerRadius = userLocationView.bounds.size.width / 2;
        userLocationView.layer.masksToBounds = YES;
        userLocationView.layer.borderColor = [UIColor whiteColor].CGColor;
        userLocationView.layer.borderWidth = 2;
        return userLocationView;
    }
    
    return nil;
}

- (void)addMovingAnnotationFrom:(CLLocation* )startLoc toCoord:(CLLocation *)destLoc {
    [self.movingAnimation reset];
    self.movingAnimation.startLoc = startLoc;
    self.movingAnimation.destLoc = destLoc;
    
    
    CLLocationDistance distance = [self.movingAnimation.startLoc distanceFromLocation:self.movingAnimation.destLoc];
    const double movingSpeed = 200.0 * 1000 / (60*60); //200公里/小时
    double duration = distance / movingSpeed;
    self.movingAnimation.duration = duration;
    self.movingAnimation.duration = 0.25; //您也可以指定一个固定动画时长
    
    [self startTimer];
}

- (void)startTimer {
    if(self.timer) {
        return;
    }
    
    //取消自动降低帧率
    self.mapView.isAllowDecreaseFrame = NO;
    __weak typeof(self) weakSelf = self;
    self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0/30 repeats:YES block:^(NSTimer * _Nonnull timer) {
        [weakSelf onTimer:timer];
    }];
}

- (void)onTimer:(NSTimer *)timer {
    self.customUserAnnotation.coordinate = [self.movingAnimation step:timer.timeInterval];
//    self.mapView.centerCoordinate = self.customUserAnnotation.coordinate;
    
    if([self.movingAnimation isFinished]) {
        //恢复自动降低帧率
        self.mapView.isAllowDecreaseFrame = YES;
        [self stopTimer];
    }
}

- (void)stopTimer {
    [self.timer invalidate];
    self.timer = nil;
}

@end

File: iOS-userLocation-smoothMove-Swift/ViewController.swift
//
//  ViewController.swift
//  iOS-userLocation-smoothMove-Swift
//
//  Created by eidan on 17/1/16.
//  Copyright © 2017年 autonavi. All rights reserved.
//

import UIKit

class ViewController: UIViewController,MAMapViewDelegate {
    
    var mapView: MAMapView!         //地图
    var customUserAnnotation: MAPointAnnotation!
    var movingAnimation: MovingAnimation!
    var timer: Timer!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.mapView = MAMapView(frame: CGRect(x: CGFloat(0), y: CGFloat(64), width: CGFloat(self.view.bounds.size.width), height: CGFloat(self.view.bounds.size.height - 64)))
        self.mapView.autoresizingMask = [UIViewAutoresizing.flexibleWidth, UIViewAutoresizing.flexibleHeight]
        self.mapView.delegate = self
        self.mapView.customizeUserLocationAccuracyCircleRepresentation = true
        self.mapView.zoomLevel = 16
        self.view.addSubview(self.mapView)
        self.view.sendSubview(toBack: self.mapView)
        
        self.customUserAnnotation = MAPointAnnotation.init()
        self.movingAnimation = MovingAnimation();
        
        
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    @IBAction func showUserLocation(_ sender: UIButton) {
        
        sender.isSelected = !sender.isSelected
        self.mapView.showsUserLocation = !self.mapView.showsUserLocation
        if self.mapView.showsUserLocation == false {
            self.mapView.removeAnnotation(self.customUserAnnotation)
        }
    }
    
    func addMovingAnnotation(from startLoc: CLLocation, toCoord destLoc: CLLocation) {
        
        self.movingAnimation.reset()
        self.movingAnimation.startLoc = startLoc
        self.movingAnimation.desLoc = destLoc
        
        let distance: CLLocationDistance = (self.movingAnimation.startLoc?.distance(from: self.movingAnimation.desLoc!))!
        let movingSpeed: Double = 200.0 * 1000 / (60 * 60)  //200公里/小时
        let duration: Double = distance / movingSpeed
        self.movingAnimation.duration = duration
//        self.movingAnimation.duration = 0.25  //您也可以指定一个固定动画时长
        
        self.startTimer()
    }
    
    func startTimer() {
        
        if (self.timer != nil) {
            return
        }
        
        self.mapView.isAllowDecreaseFrame = false  //取消自动降低帧率
        self.timer = Timer.scheduledTimer(timeInterval: 1.0 / 30, target: self, selector: #selector(self.on(_:)), userInfo: nil, repeats: true)
    }
    
    func on(_ timer: Timer) {
        self.customUserAnnotation.coordinate = self.movingAnimation.step(time: timer.timeInterval)
        if self.movingAnimation.isFinished() {
            self.mapView.isAllowDecreaseFrame = true //恢复自动降低帧率
            self.stopTimer()
        }
    }
    
    func stopTimer() {
        self.timer.invalidate()
        self.timer = nil
    }
    
    
    // MARK: - MAMapViewDelegate
    
    func mapView(_ mapView: MAMapView!, didUpdate userLocation: MAUserLocation!, updatingLocation: Bool) {
        
        if updatingLocation == false || userLocation.location == nil {
            return
        }
        
        let isContained = self.mapView.annotations.contains (where: { element in
            return element as? MAPointAnnotation == self.customUserAnnotation
        })
        
        if isContained {
            
            let preCoord: CLLocationCoordinate2D = self.customUserAnnotation.coordinate
            var curCoord: CLLocationCoordinate2D = userLocation.location.coordinate
            
            /********************************
             模拟定位飘了的情况，方便测试用的
             ********************************/
            var simulatedCount: Int = 0
            if simulatedCount < 10 {
                var delta = 0.001 + Double((arc4random() % 100) / 10000)
                var delta2 = 0.001 + Double((arc4random() % 100) / 10000)
                
                delta = delta * Double(arc4random() % 2 == 0 ? 1 : -1)
                delta2 = delta2 * Double(arc4random() % 2 == 0 ? 1 : -1)
                
                curCoord.latitude = preCoord.latitude + delta
                curCoord.longitude = preCoord.longitude + delta2
                
                simulatedCount = simulatedCount + 1
                
            }
            
            let prevLoc = CLLocation(latitude: preCoord.latitude, longitude: preCoord.longitude)
            let loc = CLLocation(latitude: curCoord.latitude, longitude: curCoord.longitude)
            let prevPoint: CGPoint = self.mapView.convert(preCoord, toPointTo: self.mapView)
            let curPoint: CGPoint = self.mapView.convert(curCoord, toPointTo: self.mapView)
            
            //超过两个像素才做动画
            if fabs(curPoint.x - prevPoint.x) > 2 || fabs(curPoint.y - prevPoint.y) > 2 {
                self.addMovingAnnotation(from: prevLoc, toCoord: loc)
            } else {
                self.customUserAnnotation.coordinate = curCoord
            }

        } else {
            self.mapView.addAnnotation(self.customUserAnnotation)
            self.customUserAnnotation.coordinate = userLocation.location.coordinate
            self.mapView.centerCoordinate = userLocation.location.coordinate
        }

    }
    
    func mapView(_ mapView: MAMapView!, rendererFor overlay: MAOverlay!) -> MAOverlayRenderer! {
        return nil
    }

    func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView! {
        
        if annotation.isKind(of: MAUserLocation.self) {
            
            let pointReuseIndetifier = "defaultUserAnnotation"
            var annotationView: MAAnnotationView? = mapView.dequeueReusableAnnotationView(withIdentifier: pointReuseIndetifier)
            
            if annotationView == nil {
                annotationView = MAAnnotationView(annotation: annotation, reuseIdentifier: pointReuseIndetifier)
                annotationView?.frame = CGRect.zero
                return annotationView
            }

        }
        
        if annotation.isKind(of: MAPointAnnotation.self) {
            
            let annotationTemp : MAPointAnnotation = annotation as! MAPointAnnotation
            
            if annotationTemp == self.customUserAnnotation {
                
                let pointReuseIndetifier = "customUserAnnotation"
                var userLocationView: MAAnnotationView? = mapView.dequeueReusableAnnotationView(withIdentifier: pointReuseIndetifier)
                
                if userLocationView == nil {
                    userLocationView = MAAnnotationView(annotation: self.customUserAnnotation, reuseIdentifier: pointReuseIndetifier)
                    userLocationView?.frame = CGRect.init(x: 0, y: 0, width: 20, height: 20)
                    userLocationView?.backgroundColor = UIColor.blue.withAlphaComponent(0.8)
                    userLocationView?.layer.cornerRadius = (userLocationView?.bounds.size.width)! / 2
                    userLocationView?.layer.masksToBounds = true
                    userLocationView?.layer.borderColor = UIColor.white.cgColor
                    userLocationView?.layer.borderWidth = 2
                    return userLocationView
                }
                
            }
            
        }
        
        return nil
    }

}

